### PR TITLE
Empty buckets

### PR DESCRIPTION
**Pull request by stugots**
This fixes the empty buckets bug in /core/events/summarize, found by Alex.

Pylint:
* master: 9.48, 10, 9.70
* this pull: 9.50, 10, 9.70

### CODE REVIEW

**Comment by lexjacobs** [comment]
pastebin url of complete results from api calls mentioned below: http://pastebin.com/pjmbrkiE

~~~~~~~~~~~~
~~~~~~~~~~~~

in my local dataset, the return for api call:
`localhost:8000/core/events/summarize/?timestamp__range={"gte":1434348772926, "lte":1434435172926}&interval=3600s`

contains a bucket at: 1434384000000 with
`{
  "image.send": 1
}`

but in the following api call:

`http://localhost:8000/core/events/summarize/?timestamp__range={"gte":1434175975042, "lte":1434435175042}&interval=10800s`

`"image.send"` is not found anywhere in the results.

Assumption: since 1434384000000 is between the {gte/lte} bounds of the 2nd api call, it should be contained within the results of both api calls.

File: goldstone/core/views.py [hunk 1]
```diff
@@ -535,41 +535,80 @@ def get(self, request, uuid, *args, **kwargs):
...
+        # for time intervals.
+        search = self._get_search(request)
 
+        # Add a top-level aggregation for types.
+        search.aggs.bucket('per_type',
+                           "terms",
+                           field="_type",
+                           min_doc_count=0)
+
+        if per_host:
```
Comment (jxstanford) [inline goldstone/core/views.py]: I would probably change the per_host param to per_type and simplify the conditional.  We should be able to look at individual hosts by using an appropriate query filter.

Comment (stugots) [inline goldstone/core/views.py]: Are you asking for the entire "if per_host" block to be deleted, and only the code in the "else" clause to remain?  In other words, don't look for per_host in the query parameters at all?

If that's the case, the serializer can be simplified too.

Please confirm.

**Comment by jxstanford** [comment]
regarding @lexjacobs ' issue, please paste the ES query generated by the code.

File: goldstone/core/views.py (hunk 1 above)
Comment (jxstanford) [inline goldstone/core/views.py]: Not exactly.  I'm asking to replace `per_host` with `per_type`, and adjust the conditional accordingly.  If someone were only looking for an event count, they could say per_type=False, and avoid the extra processing and content to get the event type aggregation inside the time bucket.

**Comment by stugots** [comment]
@jxstanford:

```


------------------

localhost:8000/core/events/summarize/?timestamp__range={"gte":1434348772926, "lte":1434435172926}&interval=3600s

> /Users/john/Documents/goldstone-server/goldstone/core/views.py(612)get()
-> serializer = self.serializer_class(search.execute().aggregations)
(Pdb) l
607  	                                               'terms',
608  	                                               field='_type',
609  	                                               min_doc_count=0)
610  	
611  	        import pdb; pdb.set_trace()
612  ->	        serializer = self.serializer_class(search.execute().aggregations)
613  	        return Response(serializer.data)
614  	
615  	
616  	# Our API documentation extracts this docstring, hence the use of markup.
617  	class EventSearchView(ElasticListAPIView):
(Pdb) pp search
<elasticsearch_dsl.search.Search object at 0x10d4ae790>
(Pdb) pp search.to_dict()
{'aggs': {'per_interval': {'aggs': {'per_type': {'terms': {'field': '_type',
                                                           'min_doc_count': 0}}},
                           'date_histogram': {'extended_bounds': {'max': 1434435172926,
                                                                  'min': 1434348772926},
                                              'field': 'timestamp',
                                              'interval': u'3600s',
                                              'min_doc_count': 0}},
          'per_type': {'terms': {'field': '_type', 'min_doc_count': 0}}},
 'query': {'range': {u'timestamp': {'gte': 1434348772926,
                                    'lte': 1434435172926}}},
 'sort': [{'timestamp': {'order': 'desc'}}]}
(Pdb) 


---------------------------------

http://localhost:8000/core/events/summarize/?timestamp__range={"gte":1434175975042, "lte":1434435175042}&interval=10800s


> /Users/john/Documents/goldstone-server/goldstone/core/views.py(612)get()
-> serializer = self.serializer_class(search.execute().aggregations)
(Pdb) l
607  	                                               'terms',
608  	                                               field='_type',
609  	                                               min_doc_count=0)
610  	
611  	        import pdb; pdb.set_trace()
612  ->	        serializer = self.serializer_class(search.execute().aggregations)
613  	        return Response(serializer.data)
614  	
615  	
616  	# Our API documentation extracts this docstring, hence the use of markup.
617  	class EventSearchView(ElasticListAPIView):
(Pdb) pp search.to_dict()
{'aggs': {'per_interval': {'aggs': {'per_type': {'terms': {'field': '_type',
                                                           'min_doc_count': 0}}},
                           'date_histogram': {'extended_bounds': {'max': 1434435175042,
                                                                  'min': 1434175975042},
                                              'field': 'timestamp',
                                              'interval': u'10800s',
                                              'min_doc_count': 0}},
          'per_type': {'terms': {'field': '_type', 'min_doc_count': 0}}},
 'query': {'range': {u'timestamp': {'gte': 1434175975042,
                                    'lte': 1434435175042}}},
 'sort': [{'timestamp': {'order': 'desc'}}]}
(Pdb) 


-----------------

```

**Comment by jxstanford** [comment]
@stugots the query looks correct to me.  @lexjacobs can you copy/paste the actual event (with timestamp) from elasticsearch-head?  Does it consistently or intermittently omit this event?

**Comment by lexjacobs** [comment]
@jxstanford is this what you were asking for?
When I repeat the queries mentioned above in postman, this image.send event is contained in the first query, but not the second.

{
    "_index": "events_2015-06-15",
    "_type": "image.send",
    "_id": "0b812844-d10b-4db2-90a7-48a23a96578f",
    "_version": 1,
    "_score": 1,
    "_source": {
        "raw": {},
        "timestamp": "2015-06-15T16:21:04.009932",
        "traits": {
            "destination_ip": "172.24.4.100",
            "user_id": "31ebe76d822a4c709772ee7f15c14c1d",
            "service": "image.localhost",
            "bytes_sent": "13200896",
            "receiver_project": "970f2364df174040862210c9185c80ce",
            "receiver_user": "e70fcebd828349ca8f1393e62ac87756",
            "image_id": "31e49bfe-7553-409b-bc91-d86e541823d4"
        }
    }
}

**Comment by stugots** [comment]
@jxstanford: Per_type implemented...  Defaults to true.

**Comment by jxstanford** [comment]
@lexjacobs yes, that was what I was looking for.  Unfortunately, it didn't lead to an obvious answer.  I'm just going to go out on a limb and guess that this falls into the category of "ES doesn't guarantee accurate results."  

Let's not hold up the show on this one, but please keep an eye out for other cases where the data looks unusual.

File: goldstone/core/views.py [hunk 2]
```diff
@@ -535,41 +535,65 @@ def get(self, request, uuid, *args, **kwargs):
...
-                          or 3m.
-             paramType: query
 
     """
 
+    AGG_FIELD = 'timestamp'
+    AGG_NAME = 'per_interval'
     serializer_class = EventAggSerializer
-    reserved_params = ['interval', 'per_host']
+    reserved_params = ['interval', 'per_host', 'per_type']
```
Comment (jxstanford) [inline goldstone/core/views.py]: still need per_host?

File: goldstone/core/views.py [hunk 3]
```diff
@@ -535,41 +535,65 @@ def get(self, request, uuid, *args, **kwargs):
...
-        interval = self.request.query_params.get('interval', '1d')
+        # Start with a basic histogram search, having a top-level aggregation
+        # for time intervals.
+        search = self._get_search(request)
 
-        data = EventData.ranged_event_agg(base_queryset, interval)
-        serializer = self.serializer_class(data)
+        # See if the request wants per-type information.
+        per_type = \
+            self.request.query_params.get('per_type', 'True')[0].upper() == 'T'
```
Comment (jxstanford) [inline goldstone/core/views.py]: per_type = 'Twelve' results in per_type = True

how about: 
ast.literal_eval(per_type.capitalize())?  Raises ValueError: malformed string if it isn't a string representation of a [Tt]rue/[Ff]alse.

File: goldstone/core/views.py (hunk 2 above)
Comment (stugots) [inline goldstone/core/views.py]: fixed

File: goldstone/core/views.py (hunk 3 above)
Comment (stugots) [inline goldstone/core/views.py]: fixed

**Comment by jxstanford** [comment]
:+1:

**Comment by lexjacobs** [comment]
:+1:

**Pull request merged**